planner/core/optimizer.go

```diff
@@ -373,10 +374,139 @@ func postOptimize(sctx sessionctx.Context, plan PhysicalPlan) PhysicalPlan {
 	mergeContinuousSelections(plan)
 	plan = eliminateUnionScanAndLock(sctx, plan)
 	plan = enableParallelApply(sctx, plan)
+	handleFineGrainedShuffle(sctx, plan)
 	checkPlanCacheable(sctx, plan)
 	return plan
 }
 
+// Only for MPP(Window<-[Sort]<-ExchangeReceiver<-ExchangeSender).
+// TiFlashFineGrainedShuffleStreamCount:
+// == 0: fine grained shuffle is disabled.
+// > 0: use TiFlashFineGrainedShuffleStreamCount as stream count.
+// < 0: use TiFlashMaxThreads as stream count when it's greater than 0. Otherwise use DefStreamCountWhenMaxThreadsNotSet.
+func handleFineGrainedShuffle(sctx sessionctx.Context, plan PhysicalPlan) {
+	streamCount := sctx.GetSessionVars().TiFlashFineGrainedShuffleStreamCount
+	if streamCount == 0 {
+		return
+	}
+	if streamCount < 0 {
+		if sctx.GetSessionVars().TiFlashMaxThreads > 0 {
+			streamCount = sctx.GetSessionVars().TiFlashMaxThreads
+		} else {
+			streamCount = variable.DefStreamCountWhenMaxThreadsNotSet
+		}
+	}
+	setupFineGrainedShuffle(uint64(streamCount), plan)
+}
+
+func setupFineGrainedShuffle(streamCount uint64, plan PhysicalPlan) {
+	if tableReader, ok := plan.(*PhysicalTableReader); ok {
+		if _, isExchangeSender := tableReader.tablePlan.(*PhysicalExchangeSender); isExchangeSender {
+			helper := fineGrainedShuffleHelper{shuffleTarget: unknown, plans: make([]*basePhysicalPlan, 1)}
+			setupFineGrainedShuffleInternal(tableReader.tablePlan, &helper, streamCount)
+		}
+	} else {
+		for _, child := range plan.Children() {
+			setupFineGrainedShuffle(streamCount, child)
+		}
+	}
+}
+
+type shuffleTarget uint8
+
+const (
+	unknown shuffleTarget = iota
+	window
+	joinBuild
+)
+
+type fineGrainedShuffleHelper struct {
+	shuffleTarget shuffleTarget
+	plans         []*basePhysicalPlan
+}
+
+func (h *fineGrainedShuffleHelper) clear() {
+	h.shuffleTarget = unknown
+	h.plans = h.plans[:0]
+}
+
+func (h *fineGrainedShuffleHelper) addTarget(t shuffleTarget, p *basePhysicalPlan) {
```

Comment:
Maybe the function name should be `updateTarget`?

Reply:
Done